src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -51,59 +50,70 @@ class Serializer implements SerializerInterface, NormalizerInterface, Denormaliz
      */
     public const EMPTY_ARRAY_AS_OBJECT = 'empty_array_as_object';
 
-    private const SCALAR_TYPES = [
-        'int' => true,
-        'bool' => true,
-        'float' => true,
-        'string' => true,
-    ];
+    private NormalizerInterface $normalizer;
+
+    private DenormalizerInterface $denormalizer;
 
     protected ChainEncoder $encoder;
 
     protected ChainDecoder $decoder;
 
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $denormalizerCache = [];
-
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $normalizerCache = [];
-
     /**
      * @param array<NormalizerInterface|DenormalizerInterface> $normalizers
      * @param array<EncoderInterface|DecoderInterface>         $encoders
      */
     public function __construct(
-        private array $normalizers = [],
+        array $normalizers = [],
         array $encoders = [],
+        ?NormalizerInterface $normalizer = null,
+        ?DenormalizerInterface $denormalizer = null,
     ) {
-        foreach ($normalizers as $normalizer) {
-            if ($normalizer instanceof SerializerAwareInterface) {
-                $normalizer->setSerializer($this);
+        if ([] !== $normalizers && (null !== $normalizer || null !== $denormalizer)) {
+            throw new InvalidArgumentException('You cannot use an array of $normalizers with a $normalizer/$denormalizer.');
+        }
+
+        $this->normalizer = $normalizer ?? new ChainNormalizer();
+        $this->denormalizer = $denormalizer ?? new ChainDenormalizer();
+
+        if ([] !== $normalizers) {
+            trigger_deprecation('symfony/serializer', '7.1', 'Passing normalizers as first argument to "%s" is deprecated, use a chain normalizer/denormalizer instead.', __METHOD__);
+        }
+
+        foreach ($normalizers as $item) {
+            if ($item instanceof SerializerAwareInterface) {
+                if (!$item instanceof NormalizerAwareInterface) {
+                    // We assume they have started to migrate
+                    trigger_deprecation('symfony/serializer', '7.1', 'Interface %s is deprecated, maybe you are interested to use %s instead.', SerializerAwareInterface::class, NormalizerAwareInterface::class);
```

Comment:
To me, it's clear that they should use it, so I'd go directly with `Interface %s is deprecated, use %s instead.`